lib/common/format.h

```diff
@@ -0,0 +1,181 @@
+  char arg[maxArgLength + 1];  // optional %s argument value
+  int offset;  // offset to error marker
+  int length;  // length of error marker
+};
+
+template<typename CHAR = char> class FormatValidator {
+public:
+  explicit FormatValidator(const CHAR *format, size_t length, IoStmtKind stmt,
+      bool warnOnNonstandardUsage, FormatError *errors, int maxErrors)
+    : format_{format}, end_{format ? format + length : nullptr}, stmt_{stmt},
+      warnOnNonstandardUsage_{warnOnNonstandardUsage}, errors_{errors},
+      maxErrors_{maxErrors}, cursor_{format - 1} {}
+
+  int Check();
+
+private:
+  enum class TokenKind {
+    None,
+    A,
+    B,
+    BN,
+    BZ,
+    D,
+    DC,
+    DP,
+    DT,
+    E,
+    EN,
+    ES,
+    EX,
+    F,
+    G,
+    I,
+    L,
+    O,
+    P,
+    RC,
+    RD,
+    RN,
+    RP,
+    RU,
+    RZ,
+    S,
+    SP,
+    SS,
+    T,
+    TL,
+    TR,
+    X,
+    Z,
+    Colon,
+    Slash,
+    Backslash,  // nonstandard: inhibit newline on output
+    Dollar,  // nonstandard: inhibit newline on output on terminals
+    Star,
+    LParen,
+    RParen,
+    Comma,
+    Point,
+    Sign,
+    UnsignedInteger,  // value in integerValue_
+    String,  // char-literal-constant or Hollerith constant
+  };
+
+  struct Token {
+    Token &set_kind(TokenKind kind) {
+      kind_ = kind;
+      return *this;
+    }
+    Token &set_offset(int offset) {
+      offset_ = offset;
+      return *this;
+    }
+    Token &set_length(int length) {
+      length_ = length;
+      return *this;
+    }
+
+    TokenKind kind() const { return kind_; }
+    int offset() const { return offset_; }
+    int length() const { return length_; }
+
+    bool IsSet() { return kind_ != TokenKind::None; }
+
+  private:
+    TokenKind kind_{TokenKind::None};
+    int offset_{0};
+    int length_{1};
+  };
+
+  void AppendError(const char *text) { AppendError(text, token_); }
+
+  void AppendError(const char *text, Token &token, const char *arg = nullptr) {
+    FormatError *error{errorCount_ ? errors_ + errorCount_ - 1 : nullptr};
+    if (errorCount_ &&
+        (suppressErrorCascade_ || errorCount_ >= maxErrors_ ||
+            error->offset == token.offset())) {
+      return;
+    }
+    error = errors_ + errorCount_;
+    error->text = text;
+    error->offset = token.offset();
+    error->length = token.length();
+    strncpy(error->arg, arg ? arg : argString_, FormatError::maxArgLength);
+    CHECK(error->arg[FormatError::maxArgLength - 1] == 0);
+    ++errorCount_;
+    suppressErrorCascade_ = true;
+  }
+
+  CHAR NextChar();
+  CHAR LookAheadChar();
+  void Advance(TokenKind);
+  void NextToken();
+
+  void check_r(bool allowed = true);
+  bool check_w(bool required = true);
+  void check_m();
+  bool check_d(bool required = true);
+  void check_e();
+
+  const CHAR *const format_;  // format text
+  const CHAR *const end_;  // one-past-last of format_ text
+  IoStmtKind stmt_;
+  bool warnOnNonstandardUsage_;
```

Comment:
How can it warn on non-standard usage when there is no way to report warnings?

Reply:
Like this (from test io10.f90):

$ cat a.f90
  write(*, '(EN,$)')
end

$ f18 -fparse-only -fdebug-semantics a.f90
$ f18 -fparse-only -fdebug-semantics -Mstandard a.f9
a.f90:1:16: error: Expected 'EN' edit descriptor 'w' value
    write(*, '(EN,$)')
                 ^
a.f90:1:17: error: Non-standard '$' edit descriptor
    write(*, '(EN,$)')
                  ^

Reply:
That's an error, not a warning.

Reply:
It's now a warning; thanks.